### PR TITLE
Support multi-byte VarInts throughout

### PR DESCRIPTION
**Pull request by prestwich**
Includes #127 and #124.

### CODE REVIEW

**Comment by prestwich** [comment]
TODO: rewrite extract input at index in all languages

File: golang/btcspv/types.go [hunk 1]
```diff
@@ -102,16 +77,19 @@ func HeaderFromRaw(raw RawHeader, height uint32) BitcoinHeader {
 	digestLE := Hash256(raw[:])
 	digestBE := ReverseHash256Endianness(digestLE)
 	prevhashLE := ExtractPrevBlockHashLE(raw)
-	prevhash := ReverseHash256Endianness(prevhashLE)
+	prevhashBE := ReverseHash256Endianness(prevhashLE)
+  merkleRootLE := ExtractMerkleRootLE(raw)
```
Comment (ErinHales) [inline golang/btcspv/types.go]: Match indentation

File: solidity/contracts/BTCUtils.sol [hunk 2]
```diff
@@ -137,20 +157,29 @@ library BTCUtils {
...
     /// @return          The input as a byte array
-    function extractInputAtIndex(bytes memory _vin, uint8 _index) internal pure returns (bytes memory) {
-        uint256 _len;
+    function extractInputAtIndex(bytes memory _vin, uint256 _index) internal pure returns (bytes memory) {
+        uint256 _varIntDataLen;
+        uint256 _nIns;
+
+        (_varIntDataLen, _nIns) = parseVarInt(_vin);
+        require(_varIntDataLen != ERR_BAD_ARG, "Read overrun during VarInt parsing");
+        require(_index <= _nIns, "Vin read overrun");
```
Comment (wadeAlexC) [inline solidity/contracts/BTCUtils.sol]: Note that these requires can be bypassed if `determineVarIntDataLength(_vin)` is zero and `_index` is zero.

File: solidity/contracts/BTCUtils.sol [hunk 3]
```diff
@@ -137,20 +157,29 @@ library BTCUtils {
...
+        uint256 _len = 0;
+        uint256 _offset = 1 + _varIntDataLen;
 
-        for (uint8 _i = 0; _i < _index; _i ++) {
+        for (uint256 _i = 0; _i < _index; _i ++) {
             _remaining = _vin.slice(_offset, _vin.length - _offset);
             _len = determineInputLength(_remaining);
+            require(_len != ERR_BAD_ARG, "Bad VarInt in scriptSig");
             _offset = _offset + _len;
         }
```
Comment (wadeAlexC) [inline solidity/contracts/BTCUtils.sol]: In this case, this for loop is skipped

File: solidity/contracts/BTCUtils.sol [hunk 4]
```diff
@@ -137,20 +157,29 @@ library BTCUtils {
...
 
         _remaining = _vin.slice(_offset, _vin.length - _offset);
         _len = determineInputLength(_remaining);
+        require(_len != ERR_BAD_ARG, "Bad VarInt in scriptSig");
         return _vin.slice(_offset, _len);
     }
```
Comment (wadeAlexC) [inline solidity/contracts/BTCUtils.sol]: And `_vin.slice(1, _vin.length - 1)` returns what may be a totally valid input

Comment (prestwich) [reply solidity/contracts/BTCUtils.sol]: Yes, this is the intended functionality. If `index` is 0, it should skip the loop.

File: solidity/contracts/BTCUtils.sol (hunk 2 above)
Comment (prestwich) [reply solidity/contracts/BTCUtils.sol]: slight bug here. 166 should be `<` instead of `<=`. same for L343 in `extractOutputAtIndex`

**Pull request merged**